### PR TITLE
feat: Download Automation API resources

### PR DESCRIPTION
**Pull request by warber**
#### What this PR does / Why we need it:
This PR adds a `Downloader` that makes use of the automation API client for downloading automation resources.
Feature flag for automation resources needs to be enabled.

#### Special notes for your reviewer:
This PR does not include special handling of escaping templating inside workflow payloads.

#### Does this PR introduce a user-facing change?

If the user enables the feature flag for automation resources, monaco will go and try to download all (supported types) of automation resources (workflows, business calendars, scheduling rules)

image
image

### CODE REVIEW

File: pkg/download/automation/automation.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/client/automation"
+	config "github.com/dynatrace/dynatrace-configuration-as-code/pkg/config/v2"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/config/v2/coordinate"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/config/v2/parameter"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/config/v2/parameter/value"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/config/v2/template"
+	v2 "github.com/dynatrace/dynatrace-configuration-as-code/pkg/project/v2"
+)
+
+var automationTypesToResources = map[config.AutomationType]automation.ResourceType{
```
Comment (warber) [inline pkg/download/automation/automation.go:32]: pkg/config and pkg/download/automation do not share the same type for resources, so we need a mapping

Comment (UnseenWizzard) [reply pkg/download/automation/automation.go:32]: this mapping is from pkg/config -> pkg/client/automation though right? As they're used right in the client calls

nitpick/maybe-bad suggestion: maybe importing `pkg/client/automation` as `automationClient` would make the distinction to this automation pkg clearer

File: pkg/download/automation/automation.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+		}
+
+		// nothing to do if there were no configs
+		// continue with next resource
+		if len(response.Results) == 0 {
+			continue
+		}
+		log.Info("Downloaded %d objects for automation resource %s", len(response.Results), string(at.Resource))
+
+		// convert and collect all objects into a slice of configs.Config values to return
```
Comment (UnseenWizzard) [inline pkg/download/automation/automation.go:80]: nit: I'm not a big fan of these "general description comments" (specifically this one and line 59) - if the functionality is self-contained and needs a description, I'd prefer if you'd pull it into a function with a speaking name. (IMO might make sense here, for line 59 I'd personally drop the comment, downloading is exactly what I expect the automation.Download method to do)

File: pkg/download/automation/automation.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		var configs []config.Config
+		for _, obj := range response.Results {
+			configId := obj.Id
+			var content string
+			if indentedData, err := json.MarshalIndent(json.RawMessage(obj.Data), "", "  "); err == nil {
```
Comment (UnseenWizzard) [inline pkg/download/automation/automation.go:85]: nit: should we pull that marshal call into an internal util? I think every downloader uses it, and they all define 2 space indent and would have to change together if this ever changes

File: cmd/monaco/download/downloaders.go [hunk 4]
```diff
@@ -17,12 +17,38 @@
...
 type downloaders []interface{}
 
+func makeDownloaders(options downloadConfigsOptions) (downloaders, error) {
+	dtClient, err := dynatrace.CreateClient(options.environmentURL, options.auth, false,
+		dtclient.WithClientRequestLimiter(concurrency.NewLimiter(environment.GetEnvValueIntLog(environment.ConcurrentRequestsEnvKey))))
+	if err != nil {
+		return nil, err
+	}
+
+	autClient := automation.NewClient(options.environmentURL, client.NewOAuthClient(context.TODO(), client.OauthCredentials{
```
Comment (UnseenWizzard) [inline cmd/monaco/download/downloaders.go:43]: There is no concurrency limiter on the automation client?

**Review by UnseenWizzard — COMMENTED** [review]
Please look into handling errors the same way as other downloaders (I might be wrong in thinking they are more lenient though!), and the use / introduction of the same Limiting for concurrent calls as the dtClient (unless handled differently in automation client, didn't yet look past the changes in this PR)

Otherwise all nitpicks :)

File: pkg/download/automation/automation.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		// nothing to do if there were no configs
+		// continue with next resource
+		if len(response.Results) == 0 {
+			continue
+		}
+		log.Info("Downloaded %d objects for automation resource %s", len(response.Results), string(at.Resource))
```
[the remark this answers is not in the document]
Comment (UnseenWizzard) [reply pkg/download/automation/automation.go:78]: Personally as a user I'd like to get an indication that it tried to download but there was nothing - but I may be wrong about this being helpful

File: cmd/monaco/download/downloaders.go (hunk 4 above)
Comment (warber) [reply cmd/monaco/download/downloaders.go:43]: i will just add it , its easy

**Comment by warber** [comment]
@UnseenWizzard I've adressed all your comments in one commit. pls take a look :v:

**Pull request merged**